File: 198.打家劫舍/rob.go
package problem198

// 动态规划:
// 由于不可以在相邻的房间闯入，所以在当前位置n房屋可以盗窃的最大值只有两种可能：
// (1)n-1房屋可盗窃的最大值
// (2)n-2房屋可盗窃的最大值 + 当前房屋的值
// 得到动态规划方程: dp[n] = MAX(dp[n-1], dp[n-2]+num)
func rob(nums []int) int {
	prevMax, curMax := 0, 0
	for _, v := range nums {
		temp := curMax
		if prevMax+v > curMax {
			curMax = prevMax + v
		}
		prevMax = temp
	}
	return curMax
}

File: 198.打家劫舍/rob.js
// dp方程为 dp[i] = Max(dp[i-2] + num, dp[i-1])
function rob(nums) {
	if (nums.length === 1) {
		return nums[0]
	}
	const dp = [nums[0], Math.max(nums[0], nums[1])]
	for (let i = 2; i < nums.length; i++) {
		dp[i] = Math.max(dp[i - 2] + nums[i], dp[i - 1])
	}
	return dp[nums.length - 1]
}

// 滚动数组优化
function rob(nums) {
	if (nums.length === 1) {
		return nums[0]
	}
	let a = nums[0]
	let b = Math.max(nums[0], nums[1])
	for (let i = 2; i < nums.length; i++) {
		[a, b] = [b, Math.max(a + nums[i], b)]
	}
	return b
}